### ISSUE TITLE
Transactions

### ISSUE THREAD

**Issue by pimeys** [prisma/tiberius#28]
We need a higher-level system for managing transactions. It is possible to trigger transactions with the normal `BEGIN/COMMIT/ROLLBACK` mechanism, but an abstraction that handles proper cleanup, committing, rolling back and so on, and offers the same query mechanisms as the `Client` would be useful.

The transaction should be started from the `Client`. When started, Rust's ownership system should prevent the usage of `Client` until the `Transaction` is dropped. The `Transaction` should provide at least the basics: `execute` and `query`, and additionally `rollback` and `commit` methods.

When the `Transaction` is dropped, it should know did the user finish it properly using the `rollback` or `commit` method, and if not, message the client to trigger a rollback before allowing more queries to be handled.

The last part can be done in many ways, and the tricky thing here is if wanting to set the `ROLLBACK` query in the `Drop` trait, which is not by design asynchronous. Transaction should have a mutable reference to the `Client`, so one way of doing this would be to set a flag in the `Client` about a dirty state, and the client (or `Connection`) takes care of triggering the `ROLLBACK` call before executing the next query.

**Comment by pimeys** [comment]
So let's write the possibilities here:

## Calling method

The transaction API could work in two ways. First is to return a transaction object:

```rust
let mut tx = client.start_transaction().await?
tx.query(..).await?;
```

This is quite straightforward and easier to abstract. Has some problems with transaction lifetimes we'll come back to later.

The second would be a closure-based approach:

```rust
client.with_transaction(|tx| async move {
    tx.query(..).await
}).await?;
```

This makes it easier to manage the transaction lifetime, but the API is harder for situations, where one wants to store the transaction somewhere or create an abstraction over the way transactions are created (nobody else uses this approach).

## Transaction API

The `Transaction` should have the same methods as a `Client` does for querying. On top of these the normal `commit` and `rollback` facilities.

It should not be possible to use a `Client` when a `Transaction` is open and in the scope, meaning it should hold a `&mut Client` inside so the compiler would catch any bad uses.

Nesting should be possible by calling `start_transaction` on an existing `Transaction`. This then requires a trait already implemented by the `Client` and a `Transaction` so we can nest the transactions, holding a reference either to a client or the parent transaction.

## Transaction lifetime

In a perfect world the user always calls either `ROLLBACK` or `COMMIT` at the end of the transaction. In real life there can be all kinds of errors that prevents us from cleaning up.

Traditionally in synchronous Rust it has been quite easy to just implement `Drop` for the transaction, where we'd block and execute `ROLLBACK` if the transaction is in an open state.

With asynchronous Rust, there needs to be a polling `Drop` for this to be easy, or we need to design the clean-up in a different way. Some options:

- mysql_async stores a reference to a connection and sets a flag if there is an open transaction. If the transaction is not being closed properly, during the next request it'll trigger a `ROLLBACK` before doing anything else. This works fine without a direct dependency to a runtime, but what if after opening the transaction you never use the connection anymore, but also keep it in the memory?
- tokio-postgres has a channel between the client and the connection, and on `Drop` sends the `ROLLBACK` through the channel to be executed in another thread. This requires the connection future to be spawned when creating a new client for the system to work. It's a valid design, but requires this exact design for the whole crate for the `Drop` to work.
- If going with the closure calling method, the function could just catch unwinding panics and errors and trigger a `ROLLBACK` before returning to the call site.
- Another possibility is to store `Arc<Mutex<Client>>` to the `Transaction`, and on Drop use the runtime's `spawn` function to run the `ROLLBACK`, cloning the client `Arc` which is `Send` and the `Mutex` giving you `Sync`. You need some feature flags and abstraction over different runtimes in the crate for this option.

Of course having an async destructor would make all of this much easier. Boats wrote a good article about the issue: https://boats.gitlab.io/blog/post/poll-drop/

**Comment by tomhoule** [comment]
> This makes it easier to manage the transaction lifetime, but the API is harder for situations, where one wants to store the transaction somewhere or create an abstraction over the way transactions are created (nobody else uses this approach).

I think [diesel does](https://docs.rs/diesel/1.4.4/diesel/connection/trait.Connection.html#method.transaction). Of course it's a different story with async APIs, since async closures aren't a thing yet.

**Comment by pimeys** [comment]
This pull request enables the basis for spawning transactions: https://github.com/prisma/tiberius/pull/55

On top of that change it should be possible to build a higher level transaction API.

**Comment by bbigras** [comment]
Any progress on this?

**Comment by pimeys** [comment]
Is there an actual need for transaction API in Tiberius? In the current state we can't guarantee an automatic `ROLLBACK` or `COMMIT` due to Rust not having a `PollDrop` trait, and Tiberius being runtime-independent would make it quite ugly to implement.

It's not that tricky to make it work by yourself:

```rust
client.simple_query("BEGIN TRAN").await?;
/// deal with your business, do not panic or error out
match result {
  Ok(_) => client.simple_query("COMMIT").await?,
  Err(_) => client.simple_query("ROLLBACK").await?,
}
```

Now adding higher level functionality on top of what we have now in Rust, especially automatic rollbacks is kind of cumbersome and if we'd not guarantee on doing a `COMMIT` or `ROLLBACK`, the API would be kind of dangerous to use.

I'm ready for suggestions how this could be done now in Tiberius, and I'm also happy if some higher level crate (such as Diesel) would implement SQL Server support with Tiberius, and would also solve the `COMMIT`/`ROLLBACK` problem with a smart pooling solution.

For now, yes transactions work and yes, you need to structure your code a bit to make them safe.

**Comment by bbigras** [comment]
Correct me if I'm wrong, but won't a transaction get a rollback if the connection is closed? If so couldn't I just use `BEGIN TRAN` and `COMMIT` and if there's an error the connection with drop and the transaction will get a rollback?

**Comment by pimeys** [comment]
Yes. But typically you want to reuse the connection between requests, so you either hold it in a mutex, or use a connection pool. Not triggering a `ROLLBACK` might lead to leaking the transaction to other requests.

**Comment by ArunGust** [comment]
How about this ?

    let res = client.execute(
        format!(
            r#"
            BEGIN TRAN
            BEGIN TRY
                INSERT INTO ARCHIVE
                SELECT * FROM XXX WHERE id = '##{}' and status = 'completed' and updateddate  < dateadd(day,-30,GETDATE());            
                DELETE FROM XX WHERE id = '6295e23b-71b9-413e-99d2-b5b677e83989' and status = 'completed' and updateddate  < dateadd(day,-30,GETDATE());            
                COMMIT TRAN;
            END TRY
            BEGIN CATCH
                SELECT ERROR_MESSAGE();
                ROLLBACK TRAN
            END CATCH
            "#, "6295e23b-71b9-413e-99d2-b5b677e83989"),
        &[],
    ).await?;
    println!("{:?}", res);

**Comment by pimeys** [comment]
This example is quite nice for the throughput due to not opening any long-running transactions. The downside is you must write your logic in T-SQL, losing some of the compile-time checks from Rust. But, throughput-wise I'd prefer this approach for sure...

**Comment by pimeys** [comment]
@esheppa has experimental branch for long-running transaction api in [this pull request](https://github.com/prisma/tiberius/pull/129). Please take a look if interested and maybe get the ball rolling again, if you'd think the api works for you.

**Comment by lex148** [comment]
Hey I know I am very late to the party/thread, but I do have some useful info to share here. 

I actually very much agree with tiberius' decision to not support transactions directly in their library. If you look at how transactions are implemented in most other rust database libraries this is almost always pushed to the connection pool level. For example in sqlx when you drop your connection their connection pool "cleans it up" before returning it to the pool. Another words, this is typically the responsibility of something upstream even if that upstream something is bundled with the library.

If you want transactions you probably want some abstraction layer that does this for you. Full disclosure here. I am the maintainer of the rust ORM [welds](https://crates.io/crates/welds). If you are working with tiberius directly you most likely don't want an ORM. That is cool they aren't for everyone. But I do want to point out welds and other crates often abstract connection/pooling so you could easily pull just that in without pulling in a full ORM. 

For example we have several projects at work that just use our connection pool crate [welds-connections](https://crates.io/crates/welds-connections). That way they aren't pull in anything heavy and can easily deal with transactions like other crates such as sqlx.
